src/category_theory/with_term.lean

```diff
@@ -0,0 +1,322 @@
+| star (of X) _ := λ f g, pempty.elim f
+| _ star (of Y) := λ f g, pempty.elim g
+| star star star := λ _ _, punit.star
+
+instance : category.{v} (with_term C) :=
+{ hom := λ X Y, hom X Y,
+  id := λ X, id _,
+  comp := λ X Y Z f g, comp f g }
+
+/-- The inclusion from `C` into `with_term C`. -/
+def incl : C ⥤ (with_term C) :=
+{ obj := of,
+  map := λ X Y f, f }
+
+instance : full (incl : C ⥤ _) :=
+{ preimage := λ X Y f, f }
+
+instance : faithful (incl : C ⥤ _) := {}
+
+/-- Map `with_term` with respect to a functor `F : C ⥤ D`. -/
+def map {D : Type*} [category D] (F : C ⥤ D) : with_term C ⥤ with_term D :=
+{ obj := λ X,
+    match X with
+    | of x := of $ F.obj x
+    | star := star
+    end,
+  map := λ X Y f,
+    match X, Y, f with
+    | of x, of y, f := F.map f
+    | of x, star, punit.star := punit.star
+    | star, star, punit.star := punit.star
+    end }
+
+instance {X : with_term C} : unique (X ⟶ star) :=
+{ default :=
+    match X with
+    | of x := punit.star
+    | star := punit.star
+    end,
+  uniq := by tidy }
+
+/-- `with_term.star` is terminal. -/
+def star_terminal : limits.is_terminal (star : with_term C) :=
+limits.is_terminal.of_unique _
+
+/-- Lift a functor `F : C ⥤ D` to `with_term C ⥤ D`, given a terminal object of `D`. -/
+@[simps]
+def lift {D : Type*} [category D] {Z : D} (hZ : limits.is_terminal Z) (F : C ⥤ D) :
+  (with_term C) ⥤ D :=
+{ obj := λ X,
+    match X with
+    | of x := F.obj x
+    | star := Z
+    end,
+  map := λ X Y f,
+    match X, Y, f with
+    | of x, of y, f := F.map f
+    | of x, star, punit.star := hZ.from _
+    | star, star, punit.star := hZ.from _
+    end,
+  map_id' := begin
+    rintros (X|X),
+    apply F.map_id,
+    apply hZ.hom_ext,
+  end,
+  map_comp' := begin
+    rintros (X|X) (Y|Y) (Z|Z) f g,
+    any_goals {apply hZ.hom_ext},
+    tidy,
+  end }
+
+/-- The isomorphism between `incl ⋙ lift _ F` with `F`. -/
+@[simps]
+def incl_lift {D : Type*} [category D] {Z : D} (hZ : limits.is_terminal Z) (F : C ⥤ D) :
+  incl ⋙ lift hZ F ≅ F :=
+{ hom := { app := λ X, 𝟙 _ },
+  inv := { app := λ X, 𝟙 _ } }
+
+/-- The uniqueness of `lift _ F`. -/
+@[simps]
+def lift_unique {D : Type*} [category D] {Z : D} (hZ : limits.is_terminal Z) (F : C ⥤ D)
+  (G : with_term C ⥤ D) (h : incl ⋙ G ≅ F) (hG : G.obj star ≅ Z) : G ≅ lift hZ F :=
+{ hom :=
+  { app := λ X,
+      match X with
+      | of x := h.hom.app x
+      | star := hG.hom
+      end,
+    naturality' := begin
+      rintros (X|X) (Y|Y) f,
+      any_goals {apply hZ.hom_ext},
+      apply h.hom.naturality,
+      cases f,
+    end },
+  inv :=
+  { app := λ X,
+      match X with
+      | of x := h.symm.hom.app x
+      | star := hG.symm.hom
+      end,
+    naturality' := begin
+      rintros (X|X) (Y|Y) f,
+      any_goals {apply (limits.is_terminal.of_iso hZ hG.symm).hom_ext},
+      apply h.symm.hom.naturality,
+      cases f,
+    end },
+  hom_inv_id' := by {ext (X|X), tidy},
+  inv_hom_id' := by {ext (X|X), tidy} }
+
```

Comment:
Do you think it's helpful to explicitly state the consequence of these universal properties, that if you already have a terminal object you are equivalent to `with_term`? (As I did for `Mat_` in #6845.) I don't think it's necessary, but it nicely fills out the API.

Reply:
I don't think this is true. E.g. consider the case where `C` is `discrete punit`. Then `with_term C` classifies arrows, whereas `C` classifies objects.

Reply:
💡

Reply:
But now that you mention this, I think it would be worthwhile to generalize lift to take any Z object of D, and a family of morphisms F.obj x to Z with x varying, satisfying the obvious commutativity condition. 

I'll make this change soon.

Reply:
Okay, this generalization is now done. 
There is now a general `lift` construction given some (possibly nonterminal/noninitial) with some data + conditions.
The old `lift` in the case with `Z` being terminal/initial is now `lift_to_terminal` resp. `lift_to_initial`.